File: Baekjoon/Sliding_Window/12891.cpp
#include <iostream>
#include <map>

using namespace std;

int S, P;
string word;
int A, C, G, T;
map<char, int> curr_dict;

bool check_valid() {
	if (curr_dict['A'] < A) return false;
	if (curr_dict['C'] < C) return false;
	if (curr_dict['G'] < G) return false;
	if (curr_dict['T'] < T) return false;
	return true;
}

int main() {
	cin >> S >> P;
	cin >> word;
	cin >> A >> C >> G >> T;

	curr_dict['A'] = 0;
	curr_dict['C'] = 0;
	curr_dict['G'] = 0;
	curr_dict['T'] = 0;

	int count = 0;

	for (int i = 0; i < P; i++) { // 초기 윈도우
		curr_dict[word[i]] += 1;
	}
	if (check_valid() == true) {
		count += 1;
	}

	for (int i = P; i < S; i++) {
		curr_dict[word[i - P]] -= 1;
		curr_dict[word[i]] += 1;
		if (check_valid() == true) {
			count += 1;
		}
	}

	cout << count << endl;

	return 0;
}

File: Baekjoon/Sliding_Window/12891.py
S, P = map(int, input().split()) # 총 길이, 부분 길이
word = list(input())
A, C, G, T = map(int, input().split()) # 각 문자의 최소 포함 개수

def check_valid():
	if curr_dict['A'] < A:
		return False
	if curr_dict['C'] < C:
		return False
	if curr_dict['G'] < G:
		return False
	if curr_dict['T'] < T:
		return False
	return True

curr_dict = {'A': 0, 'C': 0, 'G': 0, 'T': 0}
count = 0

for i in range(P): # 초기 윈도우
	curr_dict[word[i]] += 1
if check_valid() == True:
	count += 1

for i in range(P, S):
	curr_dict[word[i - P]] -= 1
	curr_dict[word[i]] += 1
	if check_valid() == True:
		count += 1

print(count)
